### PR TITLE
bookmarks

### PR DESCRIPTION
**Pull request by digorath**
## Changes proposed in this PR:

Implemeted the new bookmarks feature.

The following components were added:
* New BookmarkHeader for DetailScreen
* New Screens:
  * BookmarkScreen
  * BookmarkCategoryScreen

The BookmarkScreen can be reached through the new headerLeft on the home screen.
The navigation to the settings screen was moved to the headerRight of the bookmark screen.

The BookmarkCategoryScreen behaves similar to the Index Screen but provides no filter options.
Filtering for a provider that one has no bookmarks of caused the whole list including the filter header to disappear.

---

SVA-62

Blocked by #138

### CODE REVIEW

**Comment by donni106** [comment]
@digorath please rebase this branch onto the latest master to get all updates of version 1.4.2 and resolve conflicts

**Comment by donni106** [comment]
if i bookmark an entry, i can see it in my list. then i go into the entry from the list and remove it from the bookmarks. then i navigate back to the list and the entry is still present there. i assume that users would be confused, as they just removed it the moment before.

**Comment by donni106** [comment]
bookmarks are not working offline. without network connection the app crashed for me.

File: src/components/index.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 export * from './map';
 export * from './screens';
 
+export * from './bookmarks';
```
Comment (donni106) [inline src/components/index.js:4]: here i think was my thought to structure the folders first separated by a line break from the files, that are directly in /components. so the bookmarks line should go up, right?

File: src/config/consts.js [hunk 2]
```diff
@@ -22,19 +22,21 @@ export const consts = {
...
   MATOMO_TRACKING: {
     SCREEN_VIEW: {
-      HOME: 'Home',
-      NEWS_ITEMS: 'News',
-      EVENT_RECORDS: 'Events',
-      POINTS_OF_INTEREST: 'Points of interest',
-      TOURS: 'Tours',
-      POINTS_OF_INTEREST_AND_TOURS: 'Points of interest and tours',
-      SERVICE: 'Service',
+      BOOKMARK_CATEGORY: 'Bookmark Category',
```
Comment (donni106) [inline src/config/consts.js:25]: im englischen `category` klein

File: src/helpers/index.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
+export * from './bookmarkHelpers';
```
Comment (donni106) [inline src/helpers/index.js:1]: filename singular please to align with the others

File: src/navigation/AppStackNavigator.js [hunk 4]
```diff
@@ -11,6 +12,7 @@ import {
 import { defaultStackNavigatorConfig } from './defaultStackNavigatorConfig';
 
 import { texts } from '../config';
+import { BookmarkCategoryScreen } from '../screens/BookmarkCategoryScreen';
```
Comment (donni106) [inline src/navigation/AppStackNavigator.js:15]: from /..screens?

**Comment by donni106** [comment]
so the data marked is sorted like in the "normal" screens, right? with the published date for news for example.
it is not the order the user marked something.
it was not written to do it in another way, but there can be different opinions on sorting in a bookmark list. maybe this needs to be a filter somewhen in the future.

File: src/helpers/bookmarkHelpers.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  );
+
+  return !!bookmarks?.[key]?.find((item) => item === id);
+};
+
+export const toggleBookmark = async (itemType: string, id: string, category?: number) => {
+  const bookmarks: BookmarkList | undefined = await readFromStore(storeKey);
+  const listQueryType = getListQueryType(itemType);
+
+  // FIXME: add error handling
```
Comment (donni106) [inline src/helpers/bookmarkHelpers.ts:51]: `FIXME:` 👍  
(well, i would define it as TODO, as it is not existing yet compared to a FIXME for something not working correctly)

**Review by donni106 — CHANGES_REQUESTED** [review]
thanks. looks quite good. i encountered some problems in using, that needs to be tackled.

**Comment by digorath** [comment]
> bookmarks are not working offline. without network connection the app crashed for me.
> 
> Screenshot_20201216-123811

this should be fixed with 8d8944110b55f037579c3bdcea418e4eb29961ec

File: src/components/index.js (hunk 1 above)
Comment (digorath) [reply src/components/index.js]: you are right. changed it in 8a84e9e25ce56257a4c54ca67a967bcc32fcc310

File: src/helpers/index.js (hunk 3 above)
Comment (digorath) [reply src/helpers/index.js]: changed it in 8a84e9e25ce56257a4c54ca67a967bcc32fcc310

File: src/config/consts.js (hunk 2 above)
Comment (digorath) [reply src/config/consts.js]: changed it in 8a84e9e25ce56257a4c54ca67a967bcc32fcc310

File: src/navigation/AppStackNavigator.js (hunk 4 above)
Comment (digorath) [reply src/navigation/AppStackNavigator.js]: changed it in 8a84e9e25ce56257a4c54ca67a967bcc32fcc310

**Comment by digorath** [comment]
> so the data marked is sorted like in the "normal" screens, right? with the published date for news for example.
> it is not the order the user marked something.
> it was not written to do it in another way, but there can be different opinions on sorting in a bookmark list. maybe this needs to be a filter somewhen in the future.

for now I changed it in 8d8944110b55f037579c3bdcea418e4eb29961ec to show the most recent 3 bookmarked items in the overview screen.
the category/type screen will show them sorted with the default sorting order for now.

File: src/config/texts.js [hunk 6]
```diff
@@ -9,6 +17,9 @@ export const texts = {
     pointsOfInterest: 'Orte',
     tours: 'Touren'
   },
+  errors: {
+    noData: 'Bitte überprüfen sie ihre Internetverbindung.'
```
Comment (donni106) [inline src/config/texts.js:21]: `sie ihre` -> `Sie Ihre`

File: src/components/bookmarks/BookmarkSection.tsx [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+  if (loading) {
+    return (
+      <LoadingContainer>
+        <ActivityIndicator color={colors.accent} />
+      </LoadingContainer>
+    );
+  }
+
+  if (!data) return null;
```
Comment (donni106) [inline src/components/bookmarks/BookmarkSection.tsx:110]: I think i reach this null, when I bookmark things while i am offline. I got a white screen.

File: src/screens/BookmarkCategoryScreen.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+      <Wrapper>
+        <RegularText>{texts.bookmarks.noBookmarksinCategory}</RegularText>
+      </Wrapper>
+    );
+  }
+
+  if (!data) {
+    return (
+      <Wrapper>
+        <RegularText>{texts.errors.noData}</RegularText>
```
Comment (donni106) [inline src/screens/BookmarkCategoryScreen.js:70]: I haven't reached this while using offline.

**Comment by donni106** [comment]
> if i bookmark an entry, i can see it in my list. then i go into the entry from the list and remove it from the bookmarks. then i navigate back to the list and the entry is still present there. i assume that users would be confused, as they just removed it the moment before.

this does work as expected now.
done in https://github.com/ikuseiGmbH/smart-village-app-app/pull/146/commits/a20d08bb2a2127b124bcdd0d983f4731243bebbd @digorath?

File: src/helpers/bookmarkHelper.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+// than the other item types in every other component
+// this way the hierarchy in the storage remains flat and homogeneous
+export const getKeyFromTypeAndCategory = (itemType: string, category?: number) =>
+  category ? `${itemType}-${category}` : itemType;
+
+export const toggleBookmark = async (itemType: string, id: string, category?: number) => {
+  const bookmarks: BookmarkList | undefined = await readFromStore(BOOKMARKS_STORE_KEY);
+  const listQueryType = getListQueryType(itemType);
+
+  // FIXME: add error handling
```
Comment (donni106) [inline src/helpers/bookmarkHelper.ts:35]: what error handling is being thought of here?

**Review by donni106 — COMMENTED** [review]
It seems to be necessary to discuss behavior for offline usage.

**Comment by digorath** [comment]
> > if i bookmark an entry, i can see it in my list. then i go into the entry from the list and remove it from the bookmarks. then i navigate back to the list and the entry is still present there. i assume that users would be confused, as they just removed it the moment before.
> 
> this does work as expected now.
> done in [a20d08b](https://github.com/ikuseiGmbH/smart-village-app-app/commit/a20d08bb2a2127b124bcdd0d983f4731243bebbd) @digorath?

yes

File: src/helpers/bookmarkHelper.ts (hunk 9 above)
Comment (digorath) [reply src/helpers/bookmarkHelper.ts:35]: once we have some sort of error reporting, we can report this case. until then we can't do much about it.

File: src/components/bookmarks/BookmarkSection.tsx (hunk 7 above)
Comment (digorath) [reply src/components/bookmarks/BookmarkSection.tsx:110]: true. this is actually an annoying one to fix as an error text would potentially be shown multiple times if added here.
we could pass in an onNoData callback to the sections from the overview screen, and have it be handled there. what do you think?

File: src/config/texts.js (hunk 6 above)
Comment (digorath) [reply src/config/texts.js]: done in fc785872ca155350a23f1886ecfd91d39ae4d089

File: src/components/bookmarks/BookmarkSection.tsx (hunk 7 above)
Comment (donni106) [reply src/components/bookmarks/BookmarkSection.tsx:110]: let us call about it with more context

Comment (donni106) [reply src/components/bookmarks/BookmarkSection.tsx:110]: this place will be used for the no data callback.

**Review by donni106 — APPROVED** [review]

**Pull request merged**